libs/error-stack/src/result.rs

```diff
@@ -67,38 +71,123 @@ pub trait ResultExt {
     ///
     /// Applies [`Report::attach_printable`] on the [`Err`] variant, refer to it for more
     /// information.
-    #[must_use]
-    fn attach_printable<A>(self, attachment: A) -> Self
+    fn attach_printable<A>(
+        self,
+        attachment: A,
+    ) -> core::result::Result<Self::Ok, Report<Self::Context>>
     where
         A: fmt::Display + fmt::Debug + Send + Sync + 'static;
 
     /// Lazily adds a new printable attachment to the [`Report`] inside the [`Result`].
     ///
     /// Applies [`Report::attach_printable`] on the [`Err`] variant, refer to it for more
     /// information.
-    #[must_use]
-    fn attach_printable_lazy<A, F>(self, attachment: F) -> Self
+    fn attach_printable_lazy<A, F>(
+        self,
+        attachment: F,
+    ) -> core::result::Result<Self::Ok, Report<Self::Context>>
     where
         A: fmt::Display + fmt::Debug + Send + Sync + 'static,
         F: FnOnce() -> A;
 
     /// Changes the context of the [`Report`] inside the [`Result`].
     ///
     /// Applies [`Report::change_context`] on the [`Err`] variant, refer to it for more information.
-    fn change_context<C>(self, context: C) -> Result<Self::Ok, C>
+    fn change_context<C>(self, context: C) -> core::result::Result<Self::Ok, Report<C>>
     where
         C: Context;
 
     /// Lazily changes the context of the [`Report`] inside the [`Result`].
     ///
     /// Applies [`Report::change_context`] on the [`Err`] variant, refer to it for more information.
-    fn change_context_lazy<C, F>(self, context: F) -> Result<Self::Ok, C>
+    fn change_context_lazy<C, F>(self, context: F) -> core::result::Result<Self::Ok, Report<C>>
     where
         C: Context,
         F: FnOnce() -> C;
 }
 
-impl<T, C> ResultExt for Result<T, C> {
+impl<T, C> ResultExt for core::result::Result<T, C>
+where
+    C: Context,
+{
+    type Context = C;
+    type Ok = T;
+
+    #[track_caller]
+    fn attach<A>(self, attachment: A) -> Result<T, C>
+    where
+        A: Send + Sync + 'static,
+    {
+        match self {
+            Ok(value) => Ok(value),
+            Err(error) => Err(Report::from(error).attach(attachment)),
```

Comment:
Out of curiosity: in your mind: is `Report::new` or `Report::from` preferred? For me it's the former (`Report::new`)

Reply:
I don't really have an opinion. Maybe we should remove `new`?

Reply:
Keeping it does not hurt, that's true.

Reply:
Actually, I think `from` is more intuitive in this context as an error already exists and we create a Report from something which already exists. For simplicity, we can keep `new`, though.